File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(ROW, COL, currString, visited):
            if currString == word:
                return True
            if ROW < 0 or ROW >= len(board) or COL < 0 or COL >= len(board[0]) or len(currString) >= len(word) or visited[ROW][COL] or board[ROW][COL] != word[len(currString)]:
                return False
            newString = currString + board[ROW][COL]
            visited[ROW][COL] = True
            res =  dfs(ROW + 1, COL, newString, visited) or dfs(ROW, COL + 1, newString, visited) or dfs(ROW - 1, COL, newString, visited) or dfs(ROW, COL - 1, newString, visited)

            visited[ROW][COL] = False
            return res

        for i in range(len(board)):
            for j in range(len(board[0])):
                visited = []
                for _ in range(len(board)):
                    visited.append([False] * len(board[0]))
                if dfs(i, j, "", visited):
                    return True
        
        return False
        
        

File: 79-word-search/word-search.cpp
class Solution {
public:

    bool dfs(vector<vector<char>>& board, string word, int row, int col, int i) {
        if(i == word.length()) {
            return true;
        } 

        if (row < 0||row == board.size() || col < 0 || col == board[0].size() || board[row][col] != word[i] || board[row][col] == '#') {
            return false;
        }

        char temp = board[row][col];
        board[row][col] = '#';

        bool result = dfs(board, word, row + 1, col, i + 1) || dfs(board, word, row - 1, col, i + 1) || dfs(board, word, row, col + 1, i + 1) || dfs(board, word, row, col - 1, i + 1);

        board[row][col] = temp;
        return result;
    }

    bool exist(vector<vector<char>>& board, string word) {
        // only dfs from characters that are the same as the starting character of `word`
        // keep track of some index `i` which is the index into word.
        // on the dfs, first check if the character we are currently on for the cell is equal to the index 
        // keep a visited set, to make sure that we are not revisiting cells on a specific path -> use unordered_set
        for(int row = 0; row < board.size(); row++) {
            for(int col = 0; col < board[0].size(); col++) {
                if(dfs(board, word, row, col, 0) == true) {
                    return true;
                }
            }
        }

        return false;
    }
};